File: python/src/n0033_search_in_rotated_sorted_array.py
def search(nums: list[int], target: int) -> int:
    last = nums[-1]
    start = 0
    end = len(nums) - 1
    while start <= end:
        mid = (start + end) // 2
        if nums[mid] == target:
            return mid
        elif target <= last and nums[mid] > last:
            start = mid + 1
        elif target > last and nums[mid] <= last:
            end = mid - 1
        elif nums[mid] < target:
            start = mid + 1
        else:
            end = mid - 1
    return -1

File: rust/src/n0033_search_in_rotated_sorted_array.rs
pub fn search(nums: &[i32], target: i32) -> i32 {
    let last = *nums.last().unwrap();
    let mut start: usize = 0;
    let mut end: usize = nums.len();
    while start < end {
        let mid = start + (end - start) / 2;
        if nums[mid] == target {
            return mid as i32;
        } else if target <= last && nums[mid] > last {
            start = mid + 1;
        } else if target > last && nums[mid] < last {
            end = mid;
        } else if nums[mid] < target {
            start = mid + 1;
        } else {
            end = mid;
        }
    }
    -1
}
